### PR TITLE
Add keras QAT tutorial

### PR DESCRIPTION
**Pull request by elad-c**
Tutorial for quantizing a simple keras model with the QAT API

### CODE REVIEW

File: tutorials/example_keras_qat.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+
+import tensorflow as tf
+from keras.datasets import mnist
+from keras import Model, layers, datasets
+import model_compression_toolkit as mct
+from model_compression_toolkit.qat.keras.quantization_facade import DEFAULT_KERAS_TPC as default_tpc
```
Comment (haihabi) [inline tutorials/example_keras_qat.py:6]: Please use the official API and don't use imports from inside MCT

File: tutorials/example_keras_qat.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import numpy as np
+
+
+# init model
+num_classes = 10
```
Comment (haihabi) [inline tutorials/example_keras_qat.py:11]: Create several functions
1. Network creation
2. Dataset creation

File: tutorials/example_keras_qat.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+# Add Channels axis to data
+x_train = np.expand_dims(x_train, -1)
+x_test = np.expand_dims(x_test, -1)
+
+# convert class vectors to binary class matrices
+y_train = tf.keras.utils.to_categorical(y_train, num_classes)
+y_test = tf.keras.utils.to_categorical(y_test, num_classes)
+
+# train float model
+batch_size = 128
```
Comment (haihabi) [inline tutorials/example_keras_qat.py:44]: Put all the parameters in a single place

File: tutorials/example_keras_qat.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import model_compression_toolkit as mct
+from model_compression_toolkit.qat.keras.quantization_facade import DEFAULT_KERAS_TPC as default_tpc
+import numpy as np
+
+
+# init model
+num_classes = 10
+input_shape = (28, 28, 1)
+
+_input = layers.Input(shape=input_shape)
```
Comment (haihabi) [inline tutorials/example_keras_qat.py:14]: add if __main__:

File: tutorials/example_keras_qat.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+def gen_representative_dataset():
+    def _generator():
+        for _img in x_train:
+            yield [_img[np.newaxis, ...]]
+    return _generator().__next__
+
+
+representative_dataset = gen_representative_dataset()
+default_tpc.tp_model.default_qco.base_config.weights_n_bits = 2
```
Comment (haihabi) [inline tutorials/example_keras_qat.py:63]: Please remove those lines and let's think of a different option.
Maybe we should build an example TPC in the tutorials folder and use it.

File: tutorials/example_keras_qat.ipynb [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "b380c492-3c53-4ec1-987e-de693a1ec1d9",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
```
Comment (reuvenperetz) [inline tutorials/example_keras_qat.ipynb:41]: Remove notebook outputs

File: tutorials/example_keras_qat.ipynb [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
...;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtensorflow\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpython\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mplatform\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m self_check\n\u001b[1;32m     23\u001b[0m \u001b[38;5;66;03m# TODO(mdan): Cleanup antipattern: import for side effects.\u001b[39;00m\n\u001b[1;32m     24\u001b[0m \n\u001b[1;32m     25\u001b[0m \u001b[38;5;66;03m# Perform pre-load sanity checks in order to produce a more actionable error.\u001b[39;00m\n\u001b[0;32m---> 26\u001b[0m \u001b[43mself_check\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpreload_check\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     28\u001b[0m \u001b[38;5;66;03m# pylint: disable=wildcard-import,g-import-not-at-top,unused-import,line-too-long\u001b[39;00m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     31\u001b[0m   \u001b[38;5;66;03m# This import is expected to fail if there is an explicit shared object\u001b[39;00m\n\u001b[1;32m     32\u001b[0m   \u001b[38;5;66;03m# dependency (with_framework_lib=true), since we do not need RTLD_GLOBAL.\u001b[39;00m\n",
```
Comment (reuvenperetz) [inline tutorials/example_keras_qat.ipynb:89]: Same...

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: tutorials/example_keras_qat.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+from keras.datasets import mnist
+from keras import Model, layers, datasets
+import model_compression_toolkit as mct
+import numpy as np
+
+
+def get_tpc():
+    # Generate a TargetPlatformModel with power of two quantization with 3 bits for
+    # activations and 2 bits for weights
+    tp = mct.target_platform
```
Comment (reuvenperetz) [inline tutorials/example_keras_qat.py:12]: Please fix comment: We're generating a Keras TargetPlatformCapabilities with 3 bits...

File: tutorials/example_keras_qat.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
+
+import tensorflow as tf
+from keras.datasets import mnist
+from keras import Model, layers, datasets
+import model_compression_toolkit as mct
+import numpy as np
+
+
+def get_tpc():
```
Comment (haihabi) [inline tutorials/example_keras_qat.py:9]: Please put comment to all functions

File: tutorials/example_keras_qat.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    score = model.evaluate(x_test, y_test, verbose=0)
+    print(f"Float model test accuracy: {score[1]:02.4f}")
+
+    # prepare model for QAT with MCT and return to user for fine-tuning
+    representative_dataset = gen_representative_dataset(x_train)
+
+    qat_model, _, _ = mct.keras_quantization_aware_training_init(model,
+                                                                 representative_dataset,
+                                                                 core_config=mct.CoreConfig(n_iter=10),
+                                                                 target_platform_capabilities=get_tpc())
```
Comment (haihabi) [inline tutorials/example_keras_qat.py:108]: Please add a comment on why we use here a different TPC and how to obtain HW-specific TPC.

File: tutorials/example_keras_qat.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
+
+import tensorflow as tf
```
Comment (haihabi) [inline tutorials/example_keras_qat.py:2]: Add Licence

File: tutorials/example_keras_qat.ipynb [hunk 12]
```diff
@@ -0,0 +1,297 @@
+{
```
Comment (haihabi) [inline tutorials/example_keras_qat.ipynb:1]: Ask amir.t how to put a Licence in notebooks.

**Review by reuvenperetz — APPROVED** [review]

**Pull request merged**